app/src/main/java/com/example/bottomnav/CAN_Data.java

```diff
@@ -0,0 +1,48 @@
+package com.example.bottomnav;
+
+class CAN_Data {
+    int id;
+    int[] data;
+
+    public CAN_Data(int id, int[] data, int len) {
+        this.id = id;
+        this.data = data;
+    }
+
+    public static CAN_Data decode(String raw) {
+        int id = -1;
+        int len;
+
+        int index = -1;
+        if (raw.charAt(0) == 't' || raw.charAt(0) == 'r') {
+            id = Integer.parseInt(raw.substring(1, 4), 16);
```

Comment:
Consider adding validation here, to check that the input contains enough characters to decode the ID, and optionally the resulting decoded ID is within the 11-bit standard ID format defined by CAN.